log4j-api/src/main/java/org/apache/logging/log4j/message/ParameterFormatter.java

```diff
@@ -240,12 +244,17 @@ static void formatMessage(
             return;
         }
 
-        // Fail if there are insufficient arguments
-        if (analysis.placeholderCount > args.length) {
-            final String message = String.format(
-                    "found %d argument placeholders, but provided %d for pattern `%s`",
-                    analysis.placeholderCount, args.length, pattern);
-            throw new IllegalArgumentException(message);
+        // #2380: check if the count of placeholder is not equal to the count of arguments
+        if (analysis.placeholderCount != argCount) {
+            final int realArgCount = args.length;
+            final int noThrowableArgCount = realArgCount - ((args[realArgCount - 1] instanceof Throwable) ? 1 : 0);
```

Comment:
```suggestion
            final int noThrowableArgCount = realArgCount - ((args[argCount - 1] instanceof Throwable) ? 1 : 0);
```

Use `argCount` instead of `args.length`: the array can have an arbitrary length (it can be reused by multiple messages), that is why the method accepts `argCount` as argument.